Expose Semantics.identifier as Windows UIA AutomationId

loic-peron-inetum-public opened the pull request:
Expose Semantics.identifier as Windows AutomationId to ease application automation.

see https://github.com/flutter/flutter/issues/148763

followup to https://github.com/flutter/engine/pull/53729 and https://github.com/flutter/engine/pull/53476

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [X] I signed the [CLA].
- [X] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

loic-peron-inetum-public committed:
Propagate flutter SemanticsNode::identifier to embedder FlutterSemanticsNode2::identifier

loic-peron-inetum-public committed:
Propagate AccessibilityBridge::SemanticsNode::identifier to AXNodeData::StringAttributes[HtmlId]

chunhtai commented:
Hi @loic-peron-inetum-public , are you still working on this pr?

loic-peron-inetum-public commented:
> Hi @loic-peron-inetum-public , are you still working on this pr?

Yes I am, but I have difficulties to find a Mac to fix the problems it created with the macOS desktop target, as well as time to work on the subject due to other pressing deadline.

The windows desktop changes have not been altered since last summer, and are used internally to automate tests on the project. They still seem to fulfill the intended purpose.

I intend to fix the tests as soon as possible, but I cannot make any promise.

Some concerns have been raised with enabling UIA support (https://github.com/flutter/engine/pull/53476#discussion_r1655256327). I have not been able to treat that and would need some guidance on this specific matter, as UIA seems mandatory for this feature: I have not been able to have AutomationId exposed without enabling UIA.

loic-sharma commented:
I've converted this into a draft while it's still a work in progress. Feel free to mark this as ready for review once you're ready! :)

loic-peron-inetum-public committed:
Propagate embedder FlutterSemanticsNode2::identifier to AccessibilityBridge::SemanticsNode::identifier

loic-peron-inetum-public committed:
Fix macOS tests

pgn001 commented:
@loic-peron-inetum-public Hi, any update on this, as we also require the AutomationID.
Any suggestion on making it to run locally until this gets merged?

loic-peron-inetum-public commented:
> I've converted this into a draft while it's still a work in progress. Feel free to mark this as ready for review once you're ready! :)

I have updated the PR and removed the draft marker, and would like to have reviews from Flutter project members on this proposal.

loic-sharma commented on engine/src/flutter/shell/platform/darwin/macos/framework/Source/FlutterEngineTest.mm:
```diff
@@ -436,6 +439,7 @@ @implementation MockableFlutterEngine
   child1.text_selection_base = -1;
   child1.text_selection_extent = -1;
   child1.label = "child 1";
+  // child1.identifier = "";
```
Should this be uncommented?

```suggestion
  child1.identifier = "";
```

If no, please delete this line

loic-peron-inetum-public replied:
This is a leftover from fixing macOS tests. I removed it.

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -1629,6 +1629,8 @@ typedef struct {
   /// Describes how much space the semantics node takes up along the z-axis.
   double thickness;
+  /// A technical identifier for the node.
+  const char* identifier;
```
This file has strict ABI compatibility requirements. As a result, this needs to be added as the last member on this struct.

For more information, see the docs at the top of this file:

https://github.com/flutter/flutter/blob/6b17f99216b12c9d81f19d66cff665f8b30c5056/engine/src/flutter/shell/platform/embedder/embedder.h#L12-L50

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -1629,6 +1629,8 @@ typedef struct {
   double elevation;
   /// Describes how much space the semantics node takes up along the z-axis.
   double thickness;
+  /// A technical identifier for the node.
```
Consider copying more of the docs from `Semantics.identifier`: https://api.flutter.dev/flutter/semantics/SemanticsNode/identifier.html

```suggestion
  /// An identifier for the semantics node in native accessibility hierarchy.
  ///
  /// This value should not be exposed to the users of the app.
  ///
  /// This is usually used for UI testing with tools that work by querying the
  /// native accessibility, like UIAutomator, XCUITest, or Appium.
```

loic-peron-inetum-public replied:
Done.

loic-sharma commented on engine/src/flutter/shell/platform/windows/flutter_platform_node_delegate_windows.h:
```diff
@@ -34,6 +34,9 @@ class FlutterPlatformNodeDelegateWindows : public FlutterPlatformNodeDelegate {
       int screen_physical_pixel_y) const override;
 
+  // |ui::AXPlatformNodeDelegate|
+  std::u16string GetAuthorUniqueId() const override;
```
What uses this? I don't follow how this gets plumed through to the UIA AutomationId

loic-peron-inetum-public replied:
It is used to get the value for AutomationID: https://github.com/flutter/flutter/blob/master/engine/src/flutter/third_party/accessibility/ax/platform/ax_platform_node_win.cc#L2146

It is defined in AXPlatformNodeDelegateBase to return an empty character string, so i override it to propagate the identifier from the semantics.

loic-sharma commented:
Could we add a unit test that verifies that the UIA automation ID is set as expected?

Also, does this affect the MSAA properties? Or is the identifier completely dropped by the MSAA tree?

loic-peron-inetum-public commented:
> Could we add a unit test that verifies that the UIA automation ID is set as expected?

I will gladly add this sort of test, by I currently do not know howt to do that. Can you give me some pointers to start with this ?

> Also, does this affect the MSAA properties? Or is the identifier completely dropped by the MSAA tree?

It is my understanding that the MSAA properties do not expose this attribute, at least without IAccessibleEx. So for now it should (and by my tests it is) completely invisible without UIA.

loic-peron-inetum-public commented:
I renamed the introduced StringAttribute kHtmlId as in Chromium, and restored and wired the minimal IAccessibleEx implementation: I get the AutomationId in my test application with standard MSAA infrastructure.

loic-peron-inetum-public committed:
Restore and enable IAccessibleEx implementation

chunhtai reviewed (commented):
overall lgtm, just have some some minor comment

chunhtai commented on engine/src/flutter/third_party/accessibility/ax/ax_enums.h:
```diff
@@ -532,6 +532,7 @@ enum class StringAttribute {
   kDisplay,
   // Only present when different from parent.
   kFontFamily,
+  kHtmlId,
```
ah this is copying from chromium? We don't need to have 1 to 1 copy if the name doesn't make sense in flutter. we have long diverge from existing chromium code

loic-peron-inetum-public replied:
Would you prefer kSemanticsId ?

loic-sharma commented on engine/src/flutter/shell/platform/embedder/embedder.h:
```diff
@@ -1693,6 +1693,11 @@ typedef struct {
+  /// This value should not be exposed to the users of the app.
+  /// This is usually used for UI testing with tools that work by querying the
+  /// native accessibility, like UIAutomator, XCUITest, or Appium.
+  const char* identifier;
```
Could you update the test that verifies this value gets sent to the embedder?

For example, this Dart code sets semantic nodes' tooltip value:

https://github.com/flutter/flutter/blob/7b8fceda91d843546d08768a311edaf240470aa1/engine/src/flutter/shell/platform/embedder/fixtures/main.dart#L195

This is verified here: https://github.com/search?q=repo%3Aflutter%2Fflutter+%22ASSERT_EQ%28std%3A%3Astrncmp%28kTooltip%2C+node-%3Etooltip%2C+sizeof%28kTooltip%29+-+1%29%22&type=code

loic-peron-inetum-public commented:
> @loic-peron-inetum-public are you able to split it into 2 PRs as requested by @loic-sharma

I will make time to rework this PR in September.

loic-peron-inetum-public committed:
Add identifier check in embedder test

loic-peron-inetum-public committed:
Propagate AXNodeData::StringAttributes[HtmlId] as AXPlatformNodeDelegate::AuthorUniqueId
